### ISSUE TITLE
Allow for the publishing of Eclipse plugins

### ISSUE THREAD

**Issue by asampal** [laurentpetit/ccw#808]
It occurs to me that it might be possible to create standard Eclipse plugins relatively easily by bundling together some parts of CCW along with one or more user plugins. This might be done by selecting the appropriate user plugin projects in the Package Explorer and invoking some action. A small UI would prompt for the name of the feature that would be created and a destination directory or the option to publish directly to the Eclipse Marketplace. The needed bits would be assembled for the user and copied to this directory or uploaded to the Marketplace. Resulting plugins should be usable in any Eclipse install where the version requirements are met.

This sounds like a good candidate for a user plugin, itself.

**Comment by arichiardi** [comment]
Yep, during the discussion about plugins I was thinking exactly about this. We could export the most used/usable classes from ccw so that external plugins can reuse them and then take advantage of the built-in package manager to handle all the problems related to versioning, dependencies (if any).

Laurent, it is true that user plugins are kind of an advanced feature, but it seems (from other editors as well) that customizations are required nowadays as first class citizen.

Sure this means to go over all the classes and probably clean them up a bit in terms of abstraction if we want to expose them, but it is good and sooner or later it will be necessary.

What do you think?

**Comment by laurentpetit** [comment]
I'm all in favor of doing whatever is possible so that more users can tweak existing / write new User Plugins!

I would not set the ambitions too high at first, in terms of generality of a solution (managing dependencies, allowing native dependencies, isolating each in its own bundle, using the eclipse p2 mechanism). Because it's certainly better to start with a well defined core, with a well defined set of features, and increment from that. Better than building or abstracting too far away from actual use cases.

For instance, to answer to the specific questions:
- there's no "Eclipse Marketplace publication": the eclipse market place is just metadata. The plugins (eclipse bundles), and their associated features, mandatory plugins, companion content.xml and artifacts.xml files, must be located elsewhere.

I do not intend yet to make CCW User Plugins distributable without CCW. It's not a way to write arbitrary Eclipse User Plugins. Other people have tried that in the past (Eclipse Monkey, etc.), and failed.

The overall Eclipse Plugin discovery is not really appealing. We can do better if we don't try to rewrite it, but rather keep the scope manageable.

What I have in mind:

Some conventions to help use the Github API programmatically, with good performance:
- All user plugins are Github repositories whose name starts with `ccw-plugin-`
- User Plugins supported by the CCW Team (to some extent) are in the `ccw-ide` Github Organization
- Dynamic: User plugins can be developed dynamically. No need to restart Eclipse in 99% of normal cases.
- Isolation: User Plugins defined in their own namespaces. No dependency between user plugins. If user plugins have the need to share code, it will be interesting to consider whether this code should be integrated to CCW
- More Clojure for CCW: the more CCW is written in Clojure itself, the better for the "dynamic" above goal. And for the ability to tweak some of its parts dynamically via User Plugins.
- Documentation: clear line for User plugins creators between what parts of CCW are "internal" (and thus they depend on it at their own risks), and which parts are "API". I have started to create new stuff as subnamespaces of the `ccw.api` package. Everything inside `ccw.api` should also be properly documented. Ideally, it would be generated alongside the existing documentation, and accessible from that.
- CCW as User Plugins? This could be interesting, but is somehow in contradiction with the previous bullet point mentioning User Plugins should not depend on each other. At least, I'm in favor of separating the "java part" of CCW and the "clojure part" of CCW so that it's possible to update the "Clojure" part of CCW via Git operations. This could be fun.

But of course, I can imagine a world where people would like to have dependencies between User plugins, or add arbitrary dependencies to User Plugins.
- it is certainly possible to derive a proper OSGi bundle from a User Plugins.
- Would this bundle require `ccw.core` ? This would mean that all installed "user plugins bundles" would depend on `ccw.core`, and, in the end, have their clojure code shared in a single Clojure instance. So there would be no real isolation at runtime. Or each bundle should "start" its own clojure instance in its bundle classloader ....
- Native dependencies could be managed via OSGi fragments
- Should the Packaging as an OSGi bundle be seen as a way to package and distribute the user plugin, or for its runtime model, or both?
- How would the dependencies be declared for a User Plugin? The current state of the Art for regular Eclipse plugins is a real mess, if you ask me. Tycho is a best effort to reconcile OSGi bundle repositories and maven repositories, and it's still super hard to understand and get right. I don't want to put this hassle on the users!!!
** So maybe we could reuse the `:dependencies` format used by leiningen in a `deps.edn` file.

I really don't see all this as straightforward, and would think twice before starting heads-on on a "solution".

More hammock-time is needed.

I think the most likely actionable items are moving the stable stuff around in CCW to place them in the `ccw.api` prefix namespace (e.g. `ccw.api.eclipse` as the public part of `ccw.eclipse`, etc.). And fixing the documentation of the stuff when it's moved to `ccw.api.*`
Another actionable item would be to see how to add clojure documentation generation to maven, so that the User Plugin API documentation is generated via maven and accessible through `documentation.asciidoc` as an external link.

**Comment by arichiardi** [comment]
Yes, good points indeed, and the cleaning to```ccw.api.*``` is really necessary.

Today while changing some stuff...I broke some plugin because I did not know that ```view_helper.clj``` was used in there. There was no way for me (as a dev) to know that.

About documentation, what about ```marginalia```? I personally like, we could impose (ok strong English word here) it as doc tool for user plugins.

**Comment by laurentpetit** [comment]
Yes, view_helper has been contributed by Chas when he still believed in CCW and it is as far as I remember the companion namespace for the ReplView Java class. 




Documentation: I was talking about documenting ccw API correctly. Not necessarily imposing anything to user plugin writers. But rather discipline for ccw itself for what is exposed as API for the user plugins.




For what I have in mind, marginalia does not seem like the right tool. It's a literate programming tool. I want a doc tool for documenting APIs. The chosen tool will need to be supported out of the box by clojure-maven-plugin, or we'll have to add support for it (not necessarily hard) in ccw's fork of clojure-maven-plugin. 






—
Envoyé avec Mailbox
